packages/jest-worker/src/Farm.js

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+'use strict';
+
+import type {
+  ChildMessage,
+  QueueChildMessage,
+  WorkerInterface,
+  OnStart,
+  OnEnd,
+} from './types';
+import {CHILD_MESSAGE_CALL} from './types';
+export default class Farm {
+  _computeWorkerKey: (string, ...Array<any>) => ?string;
+  _cacheKeys: {[string]: WorkerInterface, __proto__: null};
+  _callback: Function;
+  _last: Array<QueueChildMessage>;
+  _locks: Array<boolean>;
+  _numOfWorkers: number;
+  _offset: number;
+  _queue: Array<?QueueChildMessage>;
+
+  constructor(
+    numOfWorkers: number,
+    callback: Function,
+    computeWorkerKey?: (string, ...Array<any>) => ?string,
+  ) {
+    this._callback = callback;
+    this._numOfWorkers = numOfWorkers;
+    this._cacheKeys = Object.create(null);
+    this._queue = [];
+    this._last = [];
+    this._locks = [];
+    this._offset = 0;
+    if (computeWorkerKey) {
+      this._computeWorkerKey = computeWorkerKey;
+    }
+  }
+
+  doWork(method: string, ...args: Array<any>): Promise<mixed> {
```

Comment:
I also moved this from index here so that the farm handles the worker keys internally instead of externally